### PR TITLE
Used EJ's DI version and reworked the text

### PR DESCRIPTION
**Pull request by bobthemighty**
We now introduce "pure domain model" and "edge-to-edge with DI" as achieving
different things, and use EJ's version of the DI code because it makes the
difference more obvious.

Need to update the code assuming people are happy with this

@xtaje and @seddonym is this any better? You both identified that the FCIS and DI impls were _basically the same_. This makes the difference clearer and pushes toward why you would prefer one or the other.

I do want to make a pretty diagram showing how the components interact in both versions, but I'll need to think about how to show that.

### CODE REVIEW

File: chapter_02B_abstractions.asciidoc [hunk 1]
```diff
@@ -631,71 +581,65 @@ def ej_synchronise_dirs(reader, filesystem, source_root, dest_root):
...
+<2> Each method in our `FakeFileSystem` just appends something to the list so we
+can inspect it later. This is an example of a Spy Object.
 
 
 The advantage of this approach is that your tests act on the exact same function
-that's used by your production code.  The disadvantage is that DI usually demands
-a bit more work on the part of the reader to understand what's going on.
+that's used by your production code.  The disadvantage is that we have to make
+our stateful components explicit and we have to pass them around. DHH famously
+described this as "test damage".
```
Comment (xtaje) [inline chapter_02B_abstractions.asciidoc]: I'm sure that some readers would be asking at this point, "Why bother with the fake when you could just use a unittest.mock?"  This might be a more natural spot to put the blurb about London vs Detroit style testing.

**Comment by xtaje** [comment]
> We now introduce "pure domain model" and "edge-to-edge with DI" as achieving
> different things, and use EJ's version of the DI code because it makes the
> difference more obvious.
> 
> Need to update the code assuming people are happy with this
> 
> @xtaje and @seddonym is this any better? You both identified that the FCIS and DI impls were _basically the same_. This makes the difference clearer and pushes toward why you would prefer one or the other.
> 
> I do want to make a pretty diagram showing how the components interact in both versions, but I'll need to think about how to show that.

LGTM.  The distinctions between FCIS and DI approaches will be subtle for simple use cases, and in more complex cases I think there's always some corollary pattern in the other paradigm.  Assuming this chapter is just to introduce different styles of thinking about abstractions and decoupling, It's something that could be hinted at here, then discussed in more depth in the full chapter on DI.

I like these two tweets/threads on this topic.[1](https://twitter.com/unclebobmartin/status/1141726835794939906) [2](https://twitter.com/mathiasverraes/status/1153615691397943297)

File: chapter_02B_abstractions.asciidoc [hunk 2]
```diff
@@ -522,107 +522,57 @@ Our tests now act directly on the `determine_actions()` function:
...
-demonstrating the 'pure' version of DI. In real life, it's nice to combine
-them, but perhaps it would be a clearer comparison if you showed a purer
-version of each.
-
-I also think you might need to address how the patterns you propose in this
-book relate to the alternative approaches. To me, it seems the core patterns
-are much more DI than FCIS. But you could also point out that the approaches
-can be combined - it's not either/or.
-////
-
```
Comment (xtaje) [inline chapter_02B_abstractions.asciidoc]: I'd say that the big difference between the DI and FCIS snippets is that the the DI sample has side effects interleaved with the processing logic.  The FCIS snippet is clearly separable between 3 steps: 1. input side-effects when reading the filesystem and hashing, 2. pure data processing (aka the core), then 3.  output side-effects when writing to the filesystem.

You might also be able to say something about the Filesystem type encapsulating a set of specific responsibilities that form a role, which does not exist in the FCIS example.

File: chapter_02B_abstractions.asciidoc (hunk 1 above)
Comment (hjwp) [reply chapter_02B_abstractions.asciidoc]: i was also thinking today that this chapter might be a nice place for a little sidebar saying "did you notice we didn't use any mocks?  let's talk about that."

**Comment by hjwp** [comment]
this is all good stuff. just wanted to add my perspective, as a pythonista - I don't find the difference between fcis and di to be subtle at all, they're always very easy to distinguish, because any kind of di is always unintuitive and sticks out like a sore thumb. _because_ we're so used to `mock.patch`, the immediate reaction is always "why would you _do_ that"

**Comment by hjwp** [comment]
bob and I have had long discussions about it. I think fundamentally, bob's instincts are that tests are a sufficient reason to add di, whereas I want there to be some application-side justification as well.  "ok fine but what if I don't _need_ a `--dry-run` flag? yagni!" so I think the really interesting justification for the whole "don't mock, use di" thing (which I still suspect originally came out of language limitations rather than being self-evidently a good idea) is that it forces you to think about responsibilities and abstractions. which is _probably_ a good thing

File: chapter_02B_abstractions.asciidoc [hunk 3]
```diff
@@ -631,59 +581,53 @@ def ej_synchronise_dirs(reader, filesystem, source_root, dest_root):
 [source,python]
 [role="skip"]
 ----
+class FakeFileSystem(list): #<1>
```
Comment (seddonym) [inline chapter_02B_abstractions.asciidoc]: The code here and below don't appear to be in the rendered version of the book - are they meant to be?

File: chapter_02B_abstractions.asciidoc [hunk 4]
```diff
@@ -632,6 +632,52 @@ described this as "test damage".
 In either case, we can now work on fixing all the bugs in our implementation;
 enumerating tests for all the edge cases is now much easier.
 
+==== Why not just patch it out?
```
Comment (seddonym) [inline chapter_02B_abstractions.asciidoc]: Great section.

**Comment by xtaje** [comment]
> this is all good stuff. just wanted to add my perspective, as a pythonista - I don't find the difference between fcis and di to be subtle at all, they're always very easy to distinguish, because any kind of di is always unintuitive and sticks out like a sore thumb. _because_ we're so used to `mock.patch`, the immediate reaction is always "why would you _do_ that"

I likely over-reached on that statement.  In _some_ simple cases, like injecting a policy class instance, it's not terribly different from just using a higher-order function. I also feel like there's some argument to be made about "ports-and-adapters" just being IO monads, but it would take a Haskell-ista to do it justice.

**Comment by xtaje** [comment]
> bob and I have had long discussions about it. I think fundamentally, bob's instincts are that tests are a sufficient reason to add di, whereas I want there to be some application-side justification as well. "ok fine but what if I don't _need_ a `--dry-run` flag? yagni!" so I think the really interesting justification for the whole "don't mock, use di" thing (which I still suspect originally came out of language limitations rather than being self-evidently a good idea) is that it forces you to think about responsibilities and abstractions. which is _probably_ a good thing

This a a pretty good related back-and-forth convo, if you haven't seen it before: https://martinfowler.com/articles/is-tdd-dead/.  I think testability are modularity are very closely related, but separate, '-ilities', and the comment about "language limitations" is pretty accurate.

WRT to yagni on dry-run, I'd say that the application-side justification is that, how you handle the output and/or results is very likely to change, and dry-run is the simplest example of that.

For "dont mock, use di", I think the more accurate distinction is "patching vs injection", and "mocks vs other test doubles", because you can mix-and-match (patch a fake, inject a mock, etc.).

**Comment by hjwp** [comment]
all those discussions from the rails community are fascinating.  i'm enjoying re-reading this linked thing - https://gist.github.com/dhh/4849a20d2ba89b34b201 - especially all the comments.  the ruby community went through this whole thing way before the python community did, or is about to if we have our way. be good to learn what we can from what happened back then.

**Comment by hjwp** [comment]
from https://docs.djangoproject.com/id/2.1/releases/1.11.17/

> Django 1.11.17 fixes several bugs in 1.11.16 and adds compatibility with Python 3.7.
